File: backend/app/models/audit_log.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database import Base


class AuditLog(Base):
    """수정 이력 로그"""
    __tablename__ = "audit_logs"

    id = Column(Integer, primary_key=True, index=True)

    # 대상 정보
    entity_type = Column(String(50), nullable=False)  # 'position', 'request', 'discussion'
    entity_id = Column(Integer, nullable=False, index=True)

    # 변경 정보
    action = Column(Text, nullable=False)  # 'create', 'update', 'delete', 'toggle', 매매계획 변경사항 등
    field_name = Column(String(100))  # 변경된 필드명
    old_value = Column(Text)  # 이전 값
    new_value = Column(Text)  # 새 값
    changes = Column(JSON)  # 여러 필드 변경 시 {field: {old, new}, ...}

    # 수행자 정보
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)

    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    user = relationship("User", backref="audit_logs")

File: spring-backend/backend/src/main/java/com/fundmessenger/audit/entity/AuditLog.java
package com.fundmessenger.audit.entity;

import com.fundmessenger.user.entity.User;
import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Type;

import java.time.OffsetDateTime;

@Entity
@Table(name = "audit_logs")
@Getter
@Setter
@NoArgsConstructor
public class AuditLog {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(name = "entity_type", length = 50, nullable = false)
    private String entityType;

    @Column(name = "entity_id", nullable = false)
    private Integer entityId;

    @Column(name = "action", columnDefinition = "text", nullable = false)
    private String action;

    @Column(name = "field_name", length = 100)
    private String fieldName;

    @Column(name = "old_value", columnDefinition = "text")
    private String oldValue;

    @Column(name = "new_value", columnDefinition = "text")
    private String newValue;

    @Type(JsonType.class)
    @Column(name = "changes", columnDefinition = "jsonb")
    private Object changes;

    @Column(name = "user_id", nullable = false, insertable = false, updatable = false)
    private Long userId;

    @CreationTimestamp
    @Column(name = "created_at", columnDefinition = "timestamptz")
    private OffsetDateTime createdAt;

    // --- ManyToOne Relationships ---

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;
}
